packages/consensus/src/types/utxo_set.cairo

```diff
@@ -13,60 +13,131 @@ use core::dict::Felt252Dict;
 use core::hash::{HashStateTrait, HashStateExTrait};
 use core::poseidon::PoseidonTrait;
 use super::transaction::OutPoint;
-use super::utreexo::{UtreexoState, UtreexoAccumulator};
 
 pub const TX_OUTPUT_STATUS_NONE: u8 = 0;
 pub const TX_OUTPUT_STATUS_UNSPENT: u8 = 1;
 pub const TX_OUTPUT_STATUS_SPENT: u8 = 2;
 
 #[derive(Default, Destruct)]
 pub struct UtxoSet {
-    /// Utreexo state.
-    pub utreexo_state: UtreexoState,
-    /// The leaves represent the poseidon hashes of a block's outpoints, i.e. utxos
+    /// Hashes of created UTXOs.
     pub leaves_to_add: Array<felt252>,
-    /// Hashes of UTXOs created within the current block(s).
-    /// Note that to preserve the ordering, cache has to be updated right after a
+    /// Hashes of spent (i.e. deleted) UTXOs.
+    pub leaves_to_delete: Array<felt252>,
+    /// Number of pending cached UTXOs that must be spent within the current block(s).
+    pub num_cached: i32,
+    /// Statuses of UTXOs created or spent within the current block(s).
+    /// Note that to preserve the ordering, statuses has to be updated right after a
     /// particular output is created or spent.
     pub cache: Felt252Dict<u8>,
 }
 
 #[generate_trait]
 pub impl UtxoSetImpl of UtxoSetTrait {
-    fn new(utreexo_state: UtreexoState) -> UtxoSet {
-        UtxoSet { utreexo_state, ..Default::default() }
-    }
-
     fn add(ref self: UtxoSet, output: OutPoint) -> Result<(), ByteArray> {
         let hash = PoseidonTrait::new().update_with(output).finalize();
         if output.data.cached {
             if self.cache.get(hash) != TX_OUTPUT_STATUS_NONE {
                 return Result::Err("The output has already been added");
             }
             self.cache.insert(hash, TX_OUTPUT_STATUS_UNSPENT);
+            self.num_cached += 1;
         } else {
+            // Note that it's possible to add duplicated non-cached UTXOs,
+            // because we do not check if it has already been added
             self.leaves_to_add.append(hash);
```

Comment:
@m-kus with the current approach it's possible to add duplicated non-cached UTXOs. Is it ok? Do you want me to fix it?

Reply:
Yes please